builder/comp-builder.nix

```diff
@@ -250,6 +271,35 @@ in stdenv.mkDerivation ({
     runHook postCheck
   '';
 
+  haddockPhase = ''
+    runHook preHaddock
+    ${lib.optionalString (doHaddock && (haskellLib.isLibrary componentId)) ''
+      docdir="${docdir "$doc"}"
+      mkdir -p "$docdir"
+
+      $SETUP_HS haddock \
+        "--html" \
+        ${lib.optionalString doHoogle "--hoogle"} \
+        ${lib.optionalString hyperlinkSource "--hyperlink-source"} \
+        ${lib.concatStringsSep " " component.setupHaddockFlags} \
+        || true  # some packages don't have haddock documentation
```

Comment:
Is there a way to detect this, rather than allowing proper failures to pass?

Reply:
Not easily, I don't think.

Whether the error from haddock is `haddock: No input file(s).` (means there is no Haddock markup in the sources) or `TestHaddock.hs:6:25: error: parse error on input ‘-- ^ qwerty’`, the exit code is 1 in both cases.

Reply:
@rvl Does `No input file(s)` mean that there *may* be sources, but they don't have haddocks, or that there are no sources?

Reply:
Yes, `No input file(s)` means that there are sources, but they don't have haddocks. You can see an example of this error when generating docs for the `cabal-simple` test case.

I have never tried running Haddock on a package with no sources at all.

In the case of no haddocks in the package, I think it would be simplest if the doc derivation would build successfully, but be empty.

Reply:
@rvl Actually, I think you're only getting `No input file(s)` because there is a library with no `exposed-modules` or `other-modules`. The issue is not that the Haskell code has no Haddocks, it's that Haddock is being run with no files since it doesn't do executables' sources by default, and is finding no sources in the library. [cabal2nix get's around this by exporting `doHaddock = false;` when it detects a library with no modules](https://github.com/NixOS/cabal2nix/blob/af7cf68dfc6fa94d4bdb6c5db7a0d8cb430654fd/src/Distribution/Nixpkgs/Haskell/FromCabal.hs#L155). I think allowing a build to succeed when there's a good reason for Haddock failing is worse than that.

Reply:
So we'll need to have cabal-to-nix expose the number of exported modules? That does feel a bit adhoc :-/

Reply:
@angerman I was suggesting just have `cabal-to-nix` export `doHaddock = false;`, not the number of modules. Though I do think there *might* be value in having it export the list of exposed module names; can't think of any use off the top of my head but I'm guessing it could come up.

Reply:
@ElvishJerricco I'm not on board with exporting `doHaddock = false` from cabal-to-nix. `doHaddock` should be a derived property (as you mentioned), not something that magically appears at the source of information. I'd rather export a list of exposed Module names instead of the number of exposed modules than exporting some magical doHaddock value. I guess my primary issue is that `doHaddock` is nothing that's actually part of the cabal package description.

Reply:
OK my mistake, I forgot that the cabal-simple test package had no `exposed-modules`.
I've taken out the `haddock || true` part.
So the user may need to manually disable haddock for some package dependencies, in order to build their package. This might be a pain for them. We can fix the automatic selection of `doHaddock` in another PR.

Reply:
@rvl I agree. Let's not mix this PR with one that does the auto-disabling of haddock if no haddocks are possible.